controllers/layered_image_handler.go

```diff
@@ -0,0 +1,138 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	ignTypes "github.com/coreos/ignition/v2/config/v3_2/types"
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	LayeredImageDeployCm = "layered-image-deploy-cm"
+)
+
+func (r *KataConfigOpenShiftReconciler) handleLayeredImageDeploymentFeature(state FeatureGateState) error {
+
+	// Check if MachineConfig exists and return the same without changing anything
+	mc, err := r.getExistingMachineConfig()
+	if err != nil {
+		r.Log.Info("Error in getting existing MachineConfig", "err", err)
+		return err
+	}
+
+	if mc != nil {
+		r.Log.Info("MachineConfig is already present. No changes will be done")
+		r.ExistingMc = mc
+		return nil
+	}
+
+	if state == Enabled {
+		r.Log.Info("LayeredImageDeployment feature is enabled")
+
+		cm := &corev1.ConfigMap{}
+		err := r.Client.Get(context.Background(), types.NamespacedName{
+			Name:      LayeredImageDeployCm,
+			Namespace: OperatorNamespace,
+		}, cm)
+		if err != nil {
+			r.Log.Info("Error in retrieving LayeredImageDeployment ConfigMap", "err", err)
+			return err
+		}
+
+		mc, err := r.createMachineConfigFromConfigMap(cm)
+		if err != nil {
+			r.Log.Info("Error in creating MachineConfig for LayeredImageDeployment from ConfigMap", "err", err)
+			return err
+		}
+
+		r.ImgMc = mc
+	} else {
+		r.Log.Info("LayeredImageDeployment feature is disabled. Resetting ImgMc")
+		// Reset ImgMc
+		r.ImgMc = nil
+	}
+
+	return nil
+}
+
+func (r *KataConfigOpenShiftReconciler) getExistingMachineConfig() (*mcfgv1.MachineConfig, error) {
+	r.Log.Info("Getting any existing MachineConfigs related to KataConfig")
+
+	// Retrieve the existing MachineConfig
+	// Check for label "app":r.kataConfig.Name
+	// and name "50-enable-sandboxed-containers-extension" or name "50-enable-sandboxed-containers-image"
+	mcList := &mcfgv1.MachineConfigList{}
+	err := r.Client.List(context.Background(), mcList)
+	if err != nil {
+		r.Log.Info("Error in listing MachineConfigs", "err", err)
+		return nil, err
+	}
+
+	for _, mc := range mcList.Items {
+		if mc.Labels["app"] == r.kataConfig.Name &&
+			(mc.Name == extension_mc_name || mc.Name == image_mc_name) {
```

Comment:
It seems like it's assumed the extension_mc and the image mc are mutually exclusive and used for the same purpose, in that case, i'd expect them to be similarly set from the user's POV, why the user is required to set the layered-image-deploy-cm? (extension is automatically set IIRC)

Reply:
The reason for having `layered-image-deploy-cm` to specify details like image and kernel params is because we want to enable the admins to change it as needed. So while  extension_mc and the image mc are mutually exclusive, the key difference is that image mc is configurable by the admin.

Reply:
In this case shouldn't we set default layered-image-deploy-cm with default values when this feature is enabled? instead of expecting user to set it

Reply:
We may set it going forward. We'll also provide examples in blogs when we describe using this feature for some scenarios. Right now we don't have any default image to set. 
Also note that these FGs are for enabling experiments. Eventually if we decide to enable the FG by default, then it goes in the CRD.